File: app/schemas/trading.py
"""Trading request / response schemas (Phase 2 – paper trading & Phase 3 – live)."""

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Phase 2 – Paper trading
# ---------------------------------------------------------------------------


class PaperAccountCreate(BaseModel):
    """Request body to create a new paper-trade account."""

    name: str = Field(..., max_length=100, description="Account label")
    initial_balance: Decimal = Field(
        default=Decimal("10000"), ge=100, description="Starting USDT balance"
    )


class PaperAccountOut(BaseModel):
    """Paper-trade account summary returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    initial_balance: Decimal
    cash: Decimal
    currency: str
    status: str
    created_at: datetime | None = None

    # Computed fields (filled by service)
    total_value: Decimal | None = None
    total_pnl: Decimal | None = None
    pnl_pct: Decimal | None = None


class PaperAccountListOut(BaseModel):
    """Paginated list of paper-trade accounts."""

    items: list[PaperAccountOut]
    total: int


# --- Orders ---


class PaperOrderCreate(BaseModel):
    """Request body to place a paper-trade order."""

    instrument_code: str = Field(..., max_length=20, description="e.g. BTC.US")
    order_type: str = Field(..., pattern="^(BUY|SELL)$", description="BUY or SELL")
    quantity: Decimal = Field(..., gt=0, description="Quantity in base asset")
    price: Decimal | None = Field(
        default=None, ge=0, description="Limit price; uses market price if omitted"
    )
    signal_id: int | None = Field(default=None, description="Associated signal (optional)")


class PaperOrderOut(BaseModel):
    """Paper-trade order returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    account_id: int
    instrument_code: str
    order_type: str
    price: Decimal | None = None
    quantity: Decimal
    filled_quantity: Decimal
    status: str
    reject_reason: str | None = None
    signal_id: int | None = None
    created_at: datetime | None = None
    filled_at: datetime | None = None


class PaperOrderListOut(BaseModel):
    """Paginated list of paper-trade orders."""

    items: list[PaperOrderOut]
    total: int


# --- Positions ---


class PaperPositionOut(BaseModel):
    """Paper-trade position returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    account_id: int
    instrument_code: str
    quantity: Decimal
    avg_cost: Decimal
    market_value: Decimal | None = None
    unrealized_pnl: Decimal | None = None
    realized_pnl: Decimal | None = None
    updated_at: datetime | None = None

    # Enriched fields (filled by service from ETFInfo)
    instrument_name: str | None = None
    current_price: Decimal | None = None
    pnl_pct: Decimal | None = None


# --- PnL ---


class PnLSummaryOut(BaseModel):
    """Profit / loss summary for one account."""

    account_id: int
    total_equity: Decimal
    cash: Decimal
    market_value: Decimal
    unrealized_pnl: Decimal
    realized_pnl: Decimal
    total_pnl: Decimal
    pnl_pct: Decimal | None = None
    trade_count: int = 0
    win_count: int = 0
    win_rate: Decimal | None = None


# ---------------------------------------------------------------------------
# Phase 3 – Live trading
# ---------------------------------------------------------------------------


class LiveConfigCreate(BaseModel):
    """Request body to create a live-trade configuration."""

    name: str = Field(..., max_length=100)
    api_key: str = Field(..., max_length=128, description="Binance API key")
    api_secret: str = Field(..., max_length=128, description="Binance API secret")
    is_testnet: bool = True
    max_order_value: Decimal = Field(default=Decimal("100"), ge=0)
    max_daily_loss: Decimal = Field(default=Decimal("500"), ge=0)
    max_daily_orders: int = Field(default=20, ge=1, le=1000)
    allowed_symbols: str | None = Field(default=None, description="JSON array string")


class LiveConfigUpdate(BaseModel):
    """Request body to update a live-trade configuration (partial)."""

    name: str | None = Field(default=None, max_length=100)
    is_enabled: bool | None = None
    max_order_value: Decimal | None = Field(default=None, ge=0)
    max_daily_loss: Decimal | None = Field(default=None, ge=0)
    max_daily_orders: int | None = Field(default=None, ge=1, le=1000)
    allowed_symbols: str | None = None


class LiveConfigOut(BaseModel):
    """Live-trade configuration returned by the API (secrets are NEVER returned)."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    is_testnet: bool
    is_enabled: bool
    max_order_value: Decimal
    max_daily_loss: Decimal
    max_daily_orders: int
    allowed_symbols: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class LiveOrderOut(BaseModel):
    """Live-trade order returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    config_id: int
    order_id_from_exchange: str | None = None
    instrument_code: str
    side: str
    order_type: str
    price: Decimal | None = None
    quantity: Decimal
    filled_quantity: Decimal
    status: str
    reject_reason: str | None = None
    signal_id: int | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class LivePositionOut(BaseModel):
    """Live-trade position returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    config_id: int
    instrument_code: str
    quantity: Decimal
    avg_cost: Decimal
    current_price: Decimal | None = None
    market_value: Decimal | None = None
    unrealized_pnl: Decimal | None = None
    realized_pnl: Decimal | None = None
    updated_at: datetime | None = None


class LiveAccountOut(BaseModel):
    """Binance account summary."""

    balances: list[dict]
    can_trade: bool
    account_type: str | None = None


class LiveOrderCreate(BaseModel):
    """Request body to place a live-trade order."""

    instrument_code: str = Field(..., max_length=20)
    side: str = Field(..., pattern="^(BUY|SELL)$")
    order_type: str = Field(default="LIMIT", pattern="^(LIMIT|MARKET)$")
    quantity: Decimal = Field(..., gt=0)
    price: Decimal | None = Field(default=None, ge=0, description="Required for LIMIT orders")
    signal_id: int | None = None


class RiskStatusOut(BaseModel):
    """Current risk-control status for a live-trade config."""

    config_id: int
    circuit_breaker_active: bool
    circuit_breaker_reason: str | None = None
    orders_today: int = 0
    realized_pnl_today: Decimal = Decimal("0")
    last_error: str | None = None


class RiskRuleOut(BaseModel):
    """Risk rule returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    rule_type: str
    param_key: str
    param_value: str
    is_active: bool

File: web/src/types/trading.ts
/** Paper-trade account summary. */
export interface PaperAccount {
  id: number;
  name: string;
  initial_balance: number;
  cash: number;
  currency: string;
  status: string;
  created_at: string | null;
  total_value: number | null;
  total_pnl: number | null;
  pnl_pct: number | null;
}

export interface PaperAccountListResponse {
  items: PaperAccount[];
  total: number;
}

/** Request body to create a paper-trade account. */
export interface PaperAccountCreate {
  name: string;
  initial_balance?: number;
}

/** Paper-trade order. */
export interface PaperOrder {
  id: number;
  account_id: number;
  instrument_code: string;
  order_type: 'BUY' | 'SELL';
  price: number | null;
  quantity: number;
  filled_quantity: number;
  status: string;
  reject_reason: string | null;
  signal_id: number | null;
  created_at: string | null;
  filled_at: string | null;
}

export interface PaperOrderListResponse {
  items: PaperOrder[];
  total: number;
}

/** Request body to place a paper-trade order. */
export interface PaperOrderCreate {
  instrument_code: string;
  order_type: 'BUY' | 'SELL';
  quantity: number;
  price?: number;
  signal_id?: number;
}

/** Paper-trade position (enriched with live price). */
export interface PaperPosition {
  id: number;
  account_id: number;
  instrument_code: string;
  quantity: number;
  avg_cost: number;
  market_value: number | null;
  unrealized_pnl: number | null;
  realized_pnl: number | null;
  updated_at: string | null;
  instrument_name: string | null;
  current_price: number | null;
  pnl_pct: number | null;
}

/** P&L summary for an account. */
export interface PnLSummary {
  account_id: number;
  total_equity: number;
  cash: number;
  market_value: number;
  unrealized_pnl: number;
  realized_pnl: number;
  total_pnl: number;
  pnl_pct: number | null;
  trade_count: number;
  win_count: number;
  win_rate: number | null;
}

// ---------------------------------------------------------------------------
// Phase 3 – Live trading
// ---------------------------------------------------------------------------

/** Live-trade configuration (secrets are never returned). */
export interface LiveConfig {
  id: number;
  name: string;
  is_testnet: boolean;
  is_enabled: boolean;
  max_order_value: number;
  max_daily_loss: number;
  max_daily_orders: number;
  allowed_symbols: string | null;
  created_at: string | null;
  updated_at: string | null;
}

/** Request body to create a live-trade config. */
export interface LiveConfigCreate {
  name: string;
  api_key: string;
  api_secret: string;
  is_testnet?: boolean;
  max_order_value?: number;
  max_daily_loss?: number;
  max_daily_orders?: number;
  allowed_symbols?: string;
}

/** Request body to update a live-trade config. */
export interface LiveConfigUpdate {
  name?: string;
  is_enabled?: boolean;
  max_order_value?: number;
  max_daily_loss?: number;
  max_daily_orders?: number;
  allowed_symbols?: string;
}

/** Live-trade order. */
export interface LiveOrder {
  id: number;
  config_id: number;
  order_id_from_exchange: string | null;
  instrument_code: string;
  side: 'BUY' | 'SELL';
  order_type: 'LIMIT' | 'MARKET';
  price: number | null;
  quantity: number;
  filled_quantity: number;
  status: string;
  reject_reason: string | null;
  signal_id: number | null;
  created_at: string | null;
  updated_at: string | null;
}

/** Live-trade position. */
export interface LivePosition {
  id: number;
  config_id: number;
  instrument_code: string;
  quantity: number;
  avg_cost: number;
  current_price: number | null;
  market_value: number | null;
  unrealized_pnl: number | null;
  realized_pnl: number | null;
  updated_at: string | null;
}

/** Binance account summary. */
export interface LiveAccount {
  balances: Array<{ asset: string; free: string; locked: string; total: string }>;
  can_trade: boolean;
  account_type: string | null;
}

/** Request body to place a live-trade order. */
export interface LiveOrderCreate {
  instrument_code: string;
  side: 'BUY' | 'SELL';
  order_type?: 'LIMIT' | 'MARKET';
  quantity: number;
  price?: number;
  signal_id?: number;
}

/** Risk-control status. */
export interface RiskStatus {
  config_id: number;
  circuit_breaker_active: boolean;
  circuit_breaker_reason: string | null;
  orders_today: number;
  realized_pnl_today: string;
  last_error: string | null;
}

/** Risk rule. */
export interface RiskRule {
  id: number;
  name: string;
  rule_type: string;
  param_key: string;
  param_value: string;
  is_active: boolean;
}
